File: 프로그래머스/2/92342. 양궁대회/양궁대회.cs
using System;

public class Solution {
    public int[] answer;
    public int[] ryan_info;
    public int[] apeach_info;
    public int maxDiff = -1;
    public int[] solution(int n, int[] info) {
        apeach_info = info;
        ryan_info = new int[11];
        answer = new int[11];
        dfs(n, 0, 0, 0);
        
        return maxDiff != -1 ? answer : new int[1] {-1};
    }
    private void dfs(int left, int ryanScore, int apeachScore, int idx){
        
        if(idx == 11){
            ryan_info[10] = left;
        
            int diff = ryanScore - apeachScore;
            if(apeachScore < ryanScore && maxDiff <= diff){
                
                if(maxDiff == diff && !isLowerArr(ryan_info)) 
                    return;
                
                maxDiff = diff;
                answer = (int[])ryan_info.Clone();
            }
            return;
        }
        
        int need = apeach_info[idx] +1;
        if(need <= left){
            ryan_info[idx] = need;
            dfs(left-need, ryanScore + (10 - idx), apeachScore, idx +1);
            ryan_info[idx] = 0;
        }
        
        if(apeach_info[idx] != 0)
            dfs(left, ryanScore, apeachScore + (10 - idx), idx +1);
        else
            dfs(left, ryanScore, apeachScore, idx + 1);
    }
    
    private bool isLowerArr(int[] new_arr){
        for(int i = 9; i >= 0; i--){
            if(answer[i] == new_arr[i]) continue;
            if(answer[i] < new_arr[i]) return true;
            break;
        }
        return false;
    }
}

File: 프로그래머스/2/92342. 양궁대회/양궁대회.cpp
#include <string>
#include <vector>

using namespace std;

int arr[11];
int max_diff = 0;
vector<int> answer = { -1 };
vector<int> ryan={0, 0, 0, 0, 0, 0, 0, 0, 0, 0};

int cmp(vector<int> ryan, vector<int> apeach) {
    int ryan_score = 0;
    int apeach_score = 0;

    for (int i = 0; i < 11; i++) {
        if (ryan[i] > apeach[i]) {
            ryan_score += 10 - i;
        }
        else {
            if (apeach[i] != 0) {
                apeach_score += 10 - i;
            }
        }
    }

    return ryan_score - apeach_score;
}

// 중복 조합을 이용해 라이언 배열 얻기
void shooting(int conut, int num, int draw, vector<int> ryan, vector<int> apeach) {
    if (conut == draw) {
        vector<int> tmp;
        for (int i = 0; i < draw; i++) {
            tmp.push_back(arr[i]);
        }

        for (int i = 0; i < draw; i++) {
            ryan[10 - tmp[i]]++;
        }

        int diff = cmp(ryan, apeach);

        if (diff > max_diff) {
            max_diff = diff;
            answer = ryan;
        }

        return;
    }

    for (int i = num; i < 11; i++) {
        arr[conut] = i;
        shooting(conut + 1, i, draw, ryan, apeach);
    }
}

vector<int> solution(int n, vector<int> info) {
    //vector<int> ryan(11, 0);
    vector<int> apeach = info;

    shooting(0, 0, n, ryan, apeach);

    return ryan;
}